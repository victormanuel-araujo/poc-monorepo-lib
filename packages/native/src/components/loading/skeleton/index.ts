export * from './skeleton.comp'
export { SkeletonProps } from './skeleton.types'
