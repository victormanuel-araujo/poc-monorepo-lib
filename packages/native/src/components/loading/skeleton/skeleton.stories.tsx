import React from 'react'

import { Skeleton, SkeletonProps } from '@components'
import { Meta, Story } from '@storybook/react'

export default {
  title: 'Loaders/Skeleton',
  component: Skeleton,
} as Meta

const Template: Story<SkeletonProps> = (args) => <Skeleton {...args} height={120} />

export const Default = Template.bind({})
Default.args = {}

export const RightToLeftAnimation = Template.bind({})
RightToLeftAnimation.args = {
  direction: 'rtl',
}

export const SlowSpeed = Template.bind({})
SlowSpeed.args = {
  speed: 200,
}

export const FastSpeed = Template.bind({})
SlowSpeed.args = {
  speed: 9000,
}

export const HighlightColor = Template.bind({})
HighlightColor.args = {
  highlightColor: 'primary',
  highlightShade: 'dark',
}

export const BackgroundColor = Template.bind({})
BackgroundColor.args = {
  backgroundColor: 'red',
  backgroundShade: 'lighter',
}

export const WithBorder = Template.bind({})
WithBorder.args = {
  customBorderRadius: 24,
}
