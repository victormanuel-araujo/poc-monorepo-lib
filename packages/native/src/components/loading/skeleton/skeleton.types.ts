import { ThemeColorsName, ThemeColorsShade } from '@theme'

export interface SkeletonProps {
  /**
   * The height to be applied on the skeleton component.
   * when used as string, the script will be consider as a percentage value.
   * when used as number, the script will use it as a pixel value.
   *
   * default is `undefined` - but the default height is `auto`
   */
  height: string | number
  /**
   * The width to be applied on the skeleton component.
   * when used as string, the script will be consider as a percentage value.
   * when used as number, the script will use it as a pixel value.
   *
   * @default `100%`
   */
  width?: string | number

  /**
   * Determines the background color of the skeleton component.
   * @default `gray`
   */
  backgroundColor?: ThemeColorsName
  /**
   * Determines the background shade of the skeleton component.
   * @default `lighter`
   */
  backgroundShade?: keyof ThemeColorsShade

  /**
   * Determines the highlight color of the skeleton component.
   * @default `white`
   */
  highlightColor?: ThemeColorsName
  /**
   * Determines the highlight shade of the skeleton component.
   * @default `base`
   */
  highlightShade?: keyof ThemeColorsShade

  /**
   * Determine a custom border radius to skeleton component
   * @default `0`
   */
  customBorderRadius?: number

  /**
   * Determines the animation speed in milliseconds. Use 0 to disable animation
   * @default `900`
   */
  speed?: number

  /**
   * Determines the animation direction, left to right or right to left
   * @default `ltr`
   */
  direction?: 'ltr' | 'rtl'
}
