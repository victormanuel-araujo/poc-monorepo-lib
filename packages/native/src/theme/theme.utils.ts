// utils functions to handle theme
import { get, defaultsDeep } from 'lodash'
import { Platform } from 'react-native'
import { css } from 'styled-components/native'

import { theme as _theme } from './theme'
import {
  Theme,
  ThemeColorsName,
  ThemeColorsShade,
  ThemeFontWeights,
  ThemeFontSizes,
  ThemePaddings,
  ThemeProps,
  ThemeFonts,
} from './theme.types'

/**
 * This function create an theme object.
 * Optionally, you can pass an custom theme object with custom values and it will
 * always be merged with the default theme (smash)
 *
 * @param theme - custom theme
 *
 * @example
 * ```ts
 * const theme_a = { a: 1, b: 2, c: 3 };
 * const theme_b = { a: 2 };
 *
 * const theme_c = createTheme(theme_a, theme_b);
 *
 * // the output (theme_c) is equal to:
 * // theme_c = { a: 2, b: 2, c: 3 };
 * ```
 */
export const createTheme = (theme?: DeepPartial<Theme>): Theme => {
  return defaultsDeep(theme, _theme)
}

/**
 * easy way to get an theme color with a shade and optionally an alpha channel
 *
 * @param colorName - the name of the color
 * @param colorShade - the shade of the color
 * @param alpha - an alpha value in hexadecimal pattern, that is: two character from '00' to 'FF'
 *
 * @example
 * ```ts
 * inside a styled-component
 * ${getThemeColor(...props)}
 *
 * outside a styled-component
 * getThemeColor(...props)({ theme })
 * ```
 */
export const getThemeColor =
  (colorName: ThemeColorsName, colorShade: keyof ThemeColorsShade = 'base', alpha?: string) =>
  ({ theme }): string => {
    const color = get(theme, ['colors', colorName], { base: '#000000' })
    const shade: string = get(color, [colorShade], color.base)

    return alpha && alpha.length === 2 ? `${shade}${alpha}` : shade
  }

/**
 * Get the style of a font weight.
 * Verify the platform where the component is used, and return te correct style
 * to apply the font weight
 *
 * @param weight - Font weight, ie light, regular, bold, etc.
 * @param family - Font family - only used on Android
 *
 * @example
 * ```ts
 * const Bold = styled.Text`
 *   ${getFontWeight('bold')}
 * `
 * ```
 */
export const getFontWeight =
  (weight: keyof ThemeFontWeights, family: keyof ThemeFonts = 'default') =>
  ({ theme }) => {
    /*
     * In android, is necessary to change the font family too, cause the
     * font-weight property not work correctly
     */
    if (Platform.OS === 'android') {
      /* get the font family based on "family" prop  */
      const fontFamily: string = get(theme, ['fonts', family])

      /* return the style, with font-family and font-weight defined correctly */
      return css`
        font-family: ${fontFamily}-${weight};
        font-weight: ${get(theme, ['fontWeights', weight])};
      `
    }

    /*
     * In another platforms, as IOS or Web, is not necessary to set a new font family
     * so,
     */
    return css`
      font-weight: ${theme.fontWeights[weight]};
    `
  }

/**
 * Get the correct style for a font size.
 * return a style, with the font size and the relative line-height.
 *
 * @param size - Font size name, ie sm, md, lg, etc.
 * @param rawValue - If it's true, returns the raw integer value
 *                   instead of a style.
 *
 * @example
 * ```ts
 * const Title = styled.Text`
 *   ${getFontSize('lg')};
 * `
 * ```
 *
 * @example - `theme` comes from styled-component's ThemeContext.
 * ```tsx
 *   return <Icon size={getFontSize('lg', true)({ theme })} />
 * ```
 */
export const getFontSize =
  (size: keyof ThemeFontSizes, rawValue = false) =>
  ({ theme }: ThemeProps<Theme>) => {
    if (rawValue) {
      return get(theme, ['fontSizes', size])
    }

    return css`
      font-size: ${get(theme, ['fontSizes', size])}px;
      line-height: ${get(theme, ['lineHeights', size])}px;
    `
  }

/**
 * Used to get a style of a font family and verify the platform where the
 * component is used, and return te correct style
 * to apply the font family
 *
 * @param family - the name of the family
 * @param weight - optionally an weight of the family. use it only in android
 *
 * *
 * @example
 * ```ts
 * const TitleFamily = styled.Text`
 *   ${getFontFamily('title')}
 * `
 * ```
 */
export const getFontFamily =
  (family: keyof ThemeFonts = 'default', weight: keyof ThemeFontWeights = 'medium') =>
  ({ theme }: ThemeProps<Theme>) => {
    /* get the font family based on "family" prop  */
    const fontFamily: string = get(theme, ['fonts', family])

    if (Platform.OS === 'android') {
      /* return the style, with font-family defined correctly */
      return css`
        font-family: ${fontFamily}-${weight};
      `
    }

    return css`
      font-family: ${fontFamily};
    `
  }

/**
 * Get a named padding from current theme.
 *
 * @param size - Padding size name, ie sm, md, lg, etc.
 * @param multiply - used if is necessary to multiply the padding value
 *
 * @example
 * ```ts
 * const Container = styled.View`
 *   padding: ${getPadding('md')} ${getPadding('lg')};
 * `
 * ```
 */
export const getPadding =
  (size: keyof ThemePaddings, multiply = 1) =>
  ({ theme }: ThemeProps<Theme>): string => {
    return theme.paddings[size] * multiply + 'px'
  }
