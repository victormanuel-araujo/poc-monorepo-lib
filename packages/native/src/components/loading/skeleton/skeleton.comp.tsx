import React, { useContext, useEffect, useMemo, useState } from 'react'

import { getThemeColor } from '@theme'
import { Animated, LayoutChangeEvent, useWindowDimensions, LayoutRectangle, Platform } from 'react-native'
import { ThemeContext } from 'styled-components/native'

import { GradientSkeleton, SkeletonWrapper } from './skeleton.styles'
import { SkeletonProps } from './skeleton.types'

/**
 * Skeleton component, is a loader with a loop animation that could start from left or right
 * depending on the prop `direction`
 * It's required to pass the height prop to assure the component functionality
 *
 * @example
 *
 * ```typescript
 * <Skeleton height={90} />
 * ```
 */
export const Skeleton: React.FC<SkeletonProps> = (props) => {
  /**
   * Bringing the theme from the context so that it correctly applies
   * the colors and shades received on the props
   */
  const theme = useContext(ThemeContext)

  /**
   * Destructure props to use each prop easier
   */
  const {
    height,
    width,
    backgroundColor,
    backgroundShade,
    highlightColor,
    highlightShade,
    customBorderRadius,
    direction,
    speed = 900,
  } = props

  /**
   * Using the window width to calculate where the animation should start or end
   * depending on direction prop
   * if 3 skeletons are being used simultaneously they all will start and end at the same place
   */
  const { width: D_WIDTH } = useWindowDimensions()

  /**
   * The value of the `onLayout` function of the skeleton wrapper view, this is also used to
   * calculate the start and end point of the animation with the width above
   */
  const [wrapperLayout, setWrapperLayout] = useState<LayoutRectangle>(null)

  /**
   * The animation controller state that handles the LinearGradient component animation
   */
  const [animation] = useState<Animated.Value>(new Animated.Value(0))

  /**
   * Function created to get the layout from nativeEvent prop of the skeleton wrapper view
   * It handles the `wrapperLayout` state and check if is necessary to set it
   *
   * @param event - Component layout event
   */
  const measureLayout = ({ nativeEvent }: LayoutChangeEvent) => {
    if (!!wrapperLayout && nativeEvent.layout.x === wrapperLayout.x) return null
    setWrapperLayout(nativeEvent.layout)
  }

  function _getNumberForHexadecimal(hexadecimal: string) {
    const [_, ...hexaValues] = hexadecimal.split('')
    const rgbaValues = []

    let count = 0
    while (count < hexaValues.length) {
      const decimal = parseInt(hexaValues[count], 16) * 15 + parseInt(hexaValues[count + 1], 16)
      rgbaValues.push(decimal)
      count += 2
    }

    return rgbaValues
  }

  const hexToRgba = (color: string): string => {
    const [R, G, B, A] = _getNumberForHexadecimal(color)
    return `rgba(${R}, ${G}, ${B}, ${(A || 255 / 255).toFixed(1)})`
  }

  /**
   * The animation interpolation variable that handles the animation input and output range
   * Here we are using the useMemo because it is a constant animation, will always have the same value
   * and doesn't have to get the same value each time the animation occurs
   * Also used to validade the `wrapperLayout` variable
   */
  const animationInterpolation = useMemo<Animated.AnimatedInterpolation>(() => {
    if (!wrapperLayout) return null

    return animation.interpolate({
      inputRange: [0, 1],
      outputRange:
        direction === 'ltr'
          ? [-wrapperLayout.x - 240, D_WIDTH - wrapperLayout.x]
          : [D_WIDTH - wrapperLayout.x, -wrapperLayout.x - 240],
    })
  }, [wrapperLayout])

  /**
   * Each time the Skeleton component is called, we create an animation loop to start it, or stop when it's necessary
   * We are also setting the animation value to 0 because if we don't, when the screen updates and the animation is still occurring
   * the animation gets the current value and moves from this value to 1, instead 0 to 1
   */
  useEffect(() => {
    animation.setValue(0)
    const animationLoop = Animated.loop(
      Animated.timing(animation, {
        toValue: 1,
        useNativeDriver: Platform.OS !== 'web',
        duration: speed,
      })
    )

    animationLoop.start()
    return animationLoop.stop
  }, [animation])

  /**
   * The wrapperToGradientColor is used to make it slight transparent using the `highlighColor` passed by props
   * to match prettier the gradient animation flow
   */
  const wrapperToGradientColor = hexToRgba(getThemeColor(highlightColor, highlightShade, '80')({ theme }))
  /**
   * The `gradientColor` is the main highlightColor applied with a little alpha to match the animation.
   * If the alpha was not applied, the highlight color could be too strong or confuded with the background color
   * is case it came to be white
   */
  const gradientColor = hexToRgba(getThemeColor(highlightColor, highlightShade, 'BB')({ theme }))

  return (
    <SkeletonWrapper
      testID="skeleton-comp"
      width={width}
      height={height}
      onLayout={measureLayout}
      backgroundColor={backgroundColor}
      backgroundShade={backgroundShade}
      customBorderRadius={customBorderRadius}
    >
      <GradientSkeleton
        testID="skeleton-gradient-comp"
        style={{ transform: [{ translateX: animationInterpolation || 0 }] }}
        colors={['transparent', wrapperToGradientColor, wrapperToGradientColor, gradientColor, 'transparent']}
        direction={direction}
        height={height}
        width={160}
      />
    </SkeletonWrapper>
  )
}

Skeleton.defaultProps = {
  width: '100%',
  backgroundColor: 'gray',
  backgroundShade: 'lighter',
  highlightColor: 'white',
  highlightShade: 'base',
  customBorderRadius: 0,
  direction: 'ltr',
  speed: 900,
}
