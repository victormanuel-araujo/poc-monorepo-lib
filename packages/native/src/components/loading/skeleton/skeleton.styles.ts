import { View, ViewProps } from '@components'
import { getThemeColor } from '@theme'
import { Animated } from 'react-native'
import LinearGradient from 'react-native-linear-gradient'
import styled, { css } from 'styled-components/native'

import { SkeletonProps } from './skeleton.types'

/**
 * Use React Native Animated library to create an animated version of LinearGradient component
 */
const AnimatedGradient = Animated.createAnimatedComponent(LinearGradient)

/**
 * This function handle the color and the shade passed from the props and also creates a transparent border
 *
 *
 * @param backgroundColor - the background color passed by props to the skeleton component
 * @param backgroundShade - the background shade passed by props to the skeleton component
 */
const getWrapperBackgroundColor = ({
  backgroundColor,
  backgroundShade,
}: Pick<SkeletonProps, 'backgroundColor' | 'backgroundShade'>) => {
  return css`
    border-color: ${getThemeColor(backgroundColor, backgroundShade, '00')};
    background-color: ${getThemeColor(backgroundColor, backgroundShade)};
  `
}

/* -------------------- */

export const GradientSkeleton = styled(AnimatedGradient).attrs(({ direction }: SkeletonProps) => ({
  start: { x: direction === 'ltr' ? 0 : 1, y: 1 },
  end: { x: direction === 'ltr' ? 1 : 0, y: 1 },
  locations: [0, 0.25, 0.6, 0.85, 1],
}))<Partial<SkeletonProps>>`
  position: absolute;
  width: ${({ width }) => width || 0}px;
  height: ${({ height }) => height || 'auto'}px;
  top: 0;
`
export const SkeletonWrapper = styled(View)<
  ViewProps & Pick<SkeletonProps, 'height' | 'backgroundColor' | 'backgroundShade' | 'customBorderRadius'>
>`
  height: ${({ height }) => height || 'auto'}px;
  border-radius: ${({ customBorderRadius }) => customBorderRadius}px;
  border-width: 1px;
  border-style: solid;
  overflow: hidden;
  ${getWrapperBackgroundColor};
`
