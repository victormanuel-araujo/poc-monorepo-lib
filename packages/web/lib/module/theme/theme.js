import { colors, fonts, fontSizes, fontWeights, lineHeights, paddings, units } from "./theme.values";
/*
 * in the future, if exists a second theme, just have to create a second
 * composition as same is `theme`.
 */

/**
 * Default gama theme - SMASH
 */

export const theme = {
  colors,
  fonts,
  fontSizes,
  fontWeights,
  lineHeights,
  units,
  paddings
};
//# sourceMappingURL=theme.js.map