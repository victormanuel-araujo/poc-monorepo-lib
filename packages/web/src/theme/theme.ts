import { Theme } from './theme.types'
import { colors, fonts, fontSizes, fontWeights, lineHeights, paddings, units } from './theme.values'
/*
 * in the future, if exists a second theme, just have to create a second
 * composition as same is `theme`.
 */

/**
 * Default gama theme - SMASH
 */
export const theme: Theme = {
  colors,
  fonts,
  fontSizes,
  fontWeights,
  lineHeights,
  units,
  paddings,
}
