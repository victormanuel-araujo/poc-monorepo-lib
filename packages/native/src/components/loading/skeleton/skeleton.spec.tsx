import React from 'react'

import { act } from '@testing-library/react-native'

import { renderWithTheme } from '../../../theme/theme.fixture'

import { Skeleton } from './skeleton.comp'

describe('Skeleton component', () => {
  const useStateSpy = jest.spyOn(React, 'useState')
  const setState = jest.fn()

  beforeEach(() => {
    useStateSpy.mockClear()
    jest.clearAllMocks()
  })

  it('should render the component correctly', () => {
    const skeleton = renderWithTheme(<Skeleton height={80} />)
    expect(skeleton.queryByTestId('skeleton-comp')).toBeDefined()
    expect(skeleton.toJSON()).toMatchSnapshot()
  })

  it('should have the default props applied', () => {
    expect(Skeleton.defaultProps).toMatchObject({
      direction: 'ltr',
      width: '100%',
      backgroundColor: 'gray',
      backgroundShade: 'lighter',
      highlightColor: 'white',
      highlightShade: 'base',
      customBorderRadius: 0,
      speed: 900,
    })
  })

  it('should have the width and height props correctly applied', () => {
    const skeletonWithoutWidth = renderWithTheme(<Skeleton height={80} />)
    const skeletonWithWidth = renderWithTheme(<Skeleton height={120} width={360} />)

    expect(skeletonWithoutWidth.getByTestId('skeleton-comp')).toHaveStyle({ height: 80 })
    expect(skeletonWithoutWidth.getByTestId('skeleton-gradient-comp')).toHaveStyle({ height: 80 })

    expect(skeletonWithWidth.getByTestId('skeleton-comp')).toHaveStyle({ height: 120, width: 360 })
  })

  it('should change the direction by props', () => {
    const skeletonLtr = renderWithTheme(<Skeleton height={120} />).getByTestId('skeleton-gradient-comp')
    const skeletonRtl = renderWithTheme(<Skeleton height={120} direction="rtl" />).getByTestId('skeleton-gradient-comp')

    expect(skeletonLtr).toHaveProp('start', { x: 0, y: 1 })
    expect(skeletonLtr).toHaveProp('end', { x: 1, y: 1 })

    expect(skeletonRtl).toHaveProp('start', { x: 1, y: 1 })
    expect(skeletonRtl).toHaveProp('end', { x: 0, y: 1 })
  })

  it('should set state for onLayout function trigger when different x value is passed', () => {
    useStateSpy.mockImplementation(() => [{ x: 360 }, setState])

    const skeleton = renderWithTheme(<Skeleton height={120} />)
    act(() => skeleton.toJSON().props.onLayout({ nativeEvent: { layout: { x: 260 } } }))
    act(() => skeleton.toJSON().props.onLayout({ nativeEvent: { layout: { x: 460 } } }))

    expect(setState).toBeCalledTimes(3)
  })

  it('should not set state for onLayout function trigger when same x value is passed', () => {
    useStateSpy.mockImplementation(() => [{ x: 360 }, setState])

    const skeleton = renderWithTheme(<Skeleton height={120} />)

    act(() => skeleton.toJSON().props.onLayout({ nativeEvent: { layout: { x: 260 } } }))
    act(() => {
      skeleton.toJSON().props.onLayout({ nativeEvent: { layout: { x: 260 } } })
    })

    expect(setState).toBeCalledTimes(2)
  })
})
