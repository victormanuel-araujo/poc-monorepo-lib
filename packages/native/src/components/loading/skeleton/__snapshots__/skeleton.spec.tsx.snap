// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton component should render the component correctly 1`] = `
<View
  backgroundColor="gray"
  backgroundShade="lighter"
  customBorderRadius={0}
  dir="column"
  display="flex"
  height={80}
  onLayout={[Function]}
  position="relative"
  style={
    Object {
      "backgroundColor": "#E0E0E0",
      "borderColor": "#E0E0E000",
      "borderRadius": 0,
      "borderStyle": "solid",
      "borderWidth": 1,
      "display": "flex",
      "flexDirection": "column",
      "height": 80,
      "overflow": "hidden",
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 0,
      "position": "relative",
      "width": "100%",
    }
  }
  testID="skeleton-comp"
  width="100%"
>
  <LinearGradient
    colors={
      Array [
        "transparent",
        "rgba(240, 240, 240, 120.0)",
        "rgba(240, 240, 240, 120.0)",
        "rgba(240, 240, 240, 176.0)",
        "transparent",
      ]
    }
    direction="ltr"
    end={
      Object {
        "x": 1,
        "y": 1,
      }
    }
    height={80}
    locations={
      Array [
        0,
        0.25,
        0.6,
        0.85,
        1,
      ]
    }
    start={
      Object {
        "x": 0,
        "y": 1,
      }
    }
    style={
      Object {
        "height": 80,
        "position": "absolute",
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": 160,
      }
    }
    testID="skeleton-gradient-comp"
    width={160}
  />
</View>
`;
